README.md
PetBattleAutoForfeit
=======================

**EN** — Prompts for immediate forfeit when you enter a pet battle with no available upgrades.  
**DE** — Schlägt das sofortige Aufgeben vor, wenn ein Haustierkampf ohne verfügbare Verbesserungen beginnt.  
**ES** — Sugiere a abandonar inmediatamente al entrar en un duelo de mascotas sin mejoras disponibles.  
**FR** — Propose de déclarer forfait immédiatement lorsque vous entrez dans un combat de mascottes où il n'y a pas d'améliorations disponibles.  
**PT** — Sugere que desistir imediatamente ao entrar numa batalha de mascote sem melhorias disponíveis.  
**RU** — Подсказка для немедленного отступления, когда начинается битва питомцев и нет доступных улучшений.

* [Download from Curse](https://www.curseforge.com/wow/addons/petbattleautoforfeit)
* [Download from WoWInterface](https://www.wowinterface.com/downloads/info21998-PetBattleAutoForfeit.html)


CHANGELOG.md
### Version 7.3.5.0

* Fixed a PlaySound error in the options panel

### Version 7.3.0.0

* Updated for WoW 7.3

### Version 7.1.0.0

* Updated for WoW 7.1

### Version 7.0.3.0

* Updated for WoW 7.0

### Version 6.2.0.11

* Updated for WoW 6.2
* Fixed upgrade icons sometimes showing in PVP pet battles
* Fixed the initial state of the "Enable" option checkbox

### Version 6.1.0.10

* Updated for WoW 6.1
* Improved handling of unobtainable wild pets

### Version 6.0.2.32

* Updated for WoW 6.0

### Version 5.4.7.29

* Added Portuguese translations from Ogadita

### Version 5.4.7.25

* Fixed several minor bugs in the options panel

### Version 5.4.1.20

* Updated for WoW 5.4
* Added German translations

### Version 5.3.0.19

* Updated for WoW 5.3

### Version 5.2.0.18

* Now considers the expected level loss when determining whether wild pets are upgrades
* Added an overlay icon on enemy pet unit frames indicating which ones are upgrades
* Added Russian translations from Wetxius
* Fixed library loading

### Version 5.2.0.14

* Added detection for non-capturable pets
* Added Spanish translations
* Now uses LibPetJournal-2.0 for better compatibility with other pet-related addons

### Version 5.1.0.6

* Fixed scanning when the pet journal is filtered
* Added French translations from L0relei

### Version 5.1.0.3

* First public release


Options.lua
--[[--------------------------------------------------------------------
	PetBattleAutoForfeit
	Prompts for immediate forfeit when you enter a pet battle with no available upgrades.
	Copyright (c) 2012-2018 Phanx <addons@phanx.net>. All rights reserved.
	https://www.wowinterface.com/downloads/info21998-PetBattleAutoForfeit.html
	https://www.curseforge.com/wow/addons/petbattleautoforfeit
	https://github.com/Phanx/PetBattleAutoForfeit
----------------------------------------------------------------------]]

local ADDON, private = ...
local L = private.L

local options = CreateFrame("Frame", ADDON.."Options", InterfaceOptionsFramePanelContainer)
options.name = GetAddOnMetadata(ADDON, "Title") or ADDON
options:Hide()
InterfaceOptions_AddCategory(options)

options:SetScript("OnShow", function(options)
	local title = options:CreateFontString("$parentTitle", "ARTWORK", "GameFontNormalLarge")
	title:SetPoint("TOPLEFT", 16, -16)
	title:SetText(options.name)
	options.title = title

	local subtext = options:CreateFontString("$parentSubText", "ARTWORK", "GameFontHighlightSmall")
	subtext:SetPoint("TOPLEFT", title, "BOTTOMLEFT", 0, -8)
	subtext:SetPoint("RIGHT", -32, 0)
	subtext:SetHeight(32)
	subtext:SetJustifyH("LEFT")
	subtext:SetJustifyV("TOP")
	subtext:SetText(GetAddOnMetadata(ADDON, "Notes"))
	options.subtext = subtext

	local enable = CreateFrame("CheckButton", "$parentEnable", options, "InterfaceOptionsCheckButtonTemplate")
	enable:SetPoint("TOPLEFT", subtext, "BOTTOMLEFT", -2, -8)
	enable.Text:SetText(L.Enable)
	enable.tooltipText = L.Enable_Tooltip
	enable:SetScript("OnClick", function(self)
		local checked = not not self:GetChecked()
		PlaySound(checked and SOUNDKIT.IG_MAINMENU_OPTION_CHECKBOX_ON or SOUNDKIT.IG_MAINMENU_OPTION_CHECKBOX_OFF)
		PetBattleAutoForfeitDB.enable = checked
	end)

	--------------------------------

	local quality = CreateFrame("Slider", "$parentQuality", options, "OptionsSliderTemplate")
	quality:SetPoint("TOPLEFT", enable, "BOTTOMLEFT", 6, -30)
	quality:SetWidth(250)
	quality:SetMinMaxValues(1, 4)
	quality:SetValueStep(1)

	quality.low = _G[quality:GetName().."Low"]
	quality.low:SetPoint("TOPLEFT", quality, "BOTTOMLEFT", 0, 0)
	quality.low:SetFormattedText("%s%s|r", ITEM_QUALITY_COLORS[0].hex, BATTLE_PET_BREED_QUALITY1)

	quality.high = _G[quality:GetName().."High"]
	quality.high:SetPoint("TOPRIGHT", quality, "BOTTOMRIGHT", 0, 0)
	quality.high:SetFormattedText("%s%s|r", ITEM_QUALITY_COLORS[3].hex, BATTLE_PET_BREED_QUALITY4)

	quality.text = _G[quality:GetName().."Text"]
	quality.text:SetText(L.MinQuality)
	quality.tooltipText = L.MinQuality_Tooltip

	quality.value = quality:CreateFontString(nil, "ARTWORK", "GameFontHighlight")
	quality.value:SetPoint("TOP", quality, "BOTTOM", 0, 0)

	quality:SetScript("OnValueChanged", function(self, value)
		value = floor(value + 0.5)
		self.value:SetFormattedText("%s%s|r", ITEM_QUALITY_COLORS[value-1].hex, _G["BATTLE_PET_BREED_QUALITY"..value])
		PetBattleAutoForfeitDB.minQuality = value
	end)

	--------------------------------

	local level = CreateFrame("Slider", "$parentLevel", options, "OptionsSliderTemplate")
	level:SetPoint("LEFT", quality, "RIGHT", 24, 0)
	level:SetWidth(250)
	level:SetMinMaxValues(1, 10)
	level:SetValueStep(1)

	level.low = _G[level:GetName().."Low"]
	level.low:SetPoint("TOPLEFT", level, "BOTTOMLEFT", 0, 0)
	level.low:SetText("+1")

	level.high = _G[level:GetName().."High"]
	level.high:SetPoint("TOPRIGHT", level, "BOTTOMRIGHT", 0, 0)
	level.high:SetText("+10")

	level.text = _G[level:GetName().."Text"]
	level.text:SetText(L.MinLevelUp)
	level.tooltipText = L.MinLevelUp_Tooltip

	level.value = level:CreateFontString(nil, "ARTWORK", "GameFontNormal")
	level.value:SetPoint("TOP", level, "BOTTOM", 0, 0)

	level:SetScript("OnValueChanged", function(self, value)
		value = floor(value + 0.5)
		self.value:SetFormattedText("+%d", value)
		PetBattleAutoForfeitDB.minLevelGain = value
	end)

	--------------------------------

	function options:refresh()
		enable:SetChecked(PetBattleAutoForfeitDB.enable)

		local q = PetBattleAutoForfeitDB.minQuality
		quality:SetValue(q)
		quality.value:SetFormattedText("%s%s|r", ITEM_QUALITY_COLORS[q-1].hex, _G["BATTLE_PET_BREED_QUALITY"..q])

		local d = PetBattleAutoForfeitDB.minLevelGain
		level:SetValue(d)
		level.value:SetFormattedText("+%d", d)
	end

	options:refresh()
	options:SetScript("OnShow", nil)
end)

SLASH_PBAF1 = "/pbaf"
SlashCmdList.PBAF = function(cmd)
	if options.about then
		InterfaceOptionsFrame_OpenToCategory(options.about)
	end
	InterfaceOptionsFrame_OpenToCategory(options)
end


Localization.lua
--[[--------------------------------------------------------------------
	PetBattleAutoForfeit
	Prompts for immediate forfeit when you enter a pet battle with no available upgrades.
	Copyright (c) 2012-2018 Phanx <addons@phanx.net>. All rights reserved.
	https://www.wowinterface.com/downloads/info21998-PetBattleAutoForfeit.html
	https://www.curseforge.com/wow/addons/petbattleautoforfeit
	https://github.com/Phanx/PetBattleAutoForfeit
----------------------------------------------------------------------]]

local _, private = ...
local L = { Enable = ENABLE }
private.L = L

-- THE REST OF THIS FILE IS AUTOMATICALLY GENERATED. SEE:
-- https://wow.curseforge.com/addons/petbattleautoforfeit/localization/

------------------------------------------------------------------------
-- English
------------------------------------------------------------------------

L.ClickForfeit = "Click anywhere to forfeit.|nRight-click to continue the battle anyway."
L.Enable_Tooltip = "You may want to disable the addon temporarily while you're leveling up pets."
L.MinLevelUp = "Minimum Level Gain"
L.MinLevelUp_Tooltip = "Pets at least this many levels higher than yours are considered upgrades at the same quality."
L.MinQuality = "Minimum Quality"
L.MinQuality_Tooltip = "Pets below this quality are never considered upgrades."
L.NoUpgrades = "No upgrades available!"

local CURRENT_LOCALE = GetLocale()
if CURRENT_LOCALE == "enUS" then return end

------------------------------------------------------------------------
-- German
------------------------------------------------------------------------

if CURRENT_LOCALE == "deDE" then

L.ClickForfeit = "Irgendwo klicken, um aufzugeben.|nRechtsklicken, um den Kampf trotzdem fortzusetzen."
L.Enable_Tooltip = "Vielleicht möchten Sie das Addon vorübergehend deaktivieren, während Sie die Stufen Ihrer Kampfhaustiere zu erhöhen."
L.MinLevelUp = "Mindeststufenerhöhung"
L.MinLevelUp_Tooltip = "Kampfhaustiere wenigstens diese vielen Stufen höher als Ihre gelten als Verbesserungen, wenn sie mindestens die gleiche Stufe sind."
L.MinQuality = "Mindestqualität"
L.MinQuality_Tooltip = "Kampfhaustiere unterhalb dieser Qualität gelten nie als Verbesserungen."
L.NoUpgrades = "Keine Verbesserungen sind verfügbar!"

return end

------------------------------------------------------------------------
-- Spanish
------------------------------------------------------------------------

if CURRENT_LOCALE == "esES" then

L.ClickForfeit = "Clic dondequiera para abandonar el duelo.|nClick derecho para continuar sin embargo el duelo."
L.Enable_Tooltip = "Puedes desear desactivar temporalmente el addon mientras levantar el nivel de tus mascotas de combate."
L.MinLevelUp = "Ganancia mínimo de nivel"
L.MinLevelUp_Tooltip = "Mascotas con la misma calidad que los tuyos se consideran mejoras si son al menos estos niveles más altos."
L.MinQuality = "Calidad mínima"
L.MinQuality_Tooltip = "Mascotas debajo de esta calidad nunca se consideran mejoras."
L.NoUpgrades = "No hay ningunas mejoras disponibles!"

return end

------------------------------------------------------------------------
-- Latin American Spanish
------------------------------------------------------------------------

if CURRENT_LOCALE == "esMX" then

L.ClickForfeit = "Clic dondequiera para abandonar el duelo.|nClick derecho para continuar sin embargo el duelo."
L.Enable_Tooltip = "Puedes desear desactivar temporalmente el addon mientras levantar el nivel de tus mascotas de combate."
L.MinLevelUp = "Ganancia mínimo de nivel"
L.MinLevelUp_Tooltip = "Mascotas con la misma calidad que los tuyos se consideran mejoras si son al menos estos niveles más altos."
L.MinQuality = "Calidad mínima"
L.MinQuality_Tooltip = "Mascotas debajo de esta calidad nunca se consideran mejoras."
L.NoUpgrades = "No hay ningunas mejoras disponibles!"

return end

------------------------------------------------------------------------
-- French
------------------------------------------------------------------------

if CURRENT_LOCALE == "frFR" then

L.ClickForfeit = "Cliquez n'importe où pour déclarer forfait.|nCliquez droit pour continuer quand même le combat."
L.Enable_Tooltip = "Vous pourriez vouloir désactiver temporairement l'addon tandis que vous montez vos mascottes en niveau."
L.MinLevelUp = "Gain de niveau minimum"
L.MinLevelUp_Tooltip = "Les mascottes qui ont au moins ce nombre de niveaux de plus que les vôtres sont considérées comme des améliorations à qualité égale."
L.MinQuality = "Qualité minimum"
L.MinQuality_Tooltip = "Les mascottes en-dessous de cette qualité ne sont jamais considérées comme des améliorations."
L.NoUpgrades = "Pas d'améliorations disponibles !"

return end

------------------------------------------------------------------------
-- Italian
------------------------------------------------------------------------

if CURRENT_LOCALE == "itIT" then

return end

------------------------------------------------------------------------
-- Brazilian Portuguese
------------------------------------------------------------------------

if CURRENT_LOCALE == "ptBR" then

L.ClickForfeit = "Clique em qualquer lugar para desistir.|nClique com botão direito para continuar a batalha."
L.Enable_Tooltip = "Você pode desativar o addon temporariamente enquanto você está nivelando seus mascotes."
L.MinLevelUp = "Ganhou o level minimo"
L.MinLevelUp_Tooltip = "Mascotes com níveis mais elevados que o seu são consideradas atualizações e com a mesma qualidade."
L.MinQuality = "Qualidade mínima"
L.MinQuality_Tooltip = "Mascotes abaixo desse nivel nunca serão consideradas atualizações."
L.NoUpgrades = "Não há atualizações disponíveis!"

return end

------------------------------------------------------------------------
-- Russian
------------------------------------------------------------------------

if CURRENT_LOCALE == "ruRU" then

L.ClickForfeit = "Кликните где-нибудь, чтобы сдаться.|nПравый-клик для продолжения битвы."
L.Enable_Tooltip = "Вы можете временно отключить аддон, пока вы повышаете уровень питомцев."
L.MinLevelUp = "Минимальный получаемый уровень"
L.MinLevelUp_Tooltip = "Питомцы, которые выше уровнем, считаются улучшением с тем же качеством."
L.MinQuality = "Минимальное качество"
L.MinQuality_Tooltip = "Питомцы ниже этого качества не считаются улучшением."
L.NoUpgrades = "Нет улучшений!"

return end

------------------------------------------------------------------------
-- Korean
------------------------------------------------------------------------

if CURRENT_LOCALE == "koKR" then

return end

------------------------------------------------------------------------
-- Simplified Chinese
------------------------------------------------------------------------

if CURRENT_LOCALE == "zhCN" then

return end

------------------------------------------------------------------------
-- Traditional Chinese
------------------------------------------------------------------------

if CURRENT_LOCALE == "zhTW" then

return end


Addon.lua
--[[--------------------------------------------------------------------
	PetBattleAutoForfeit
	Prompts for immediate forfeit when you enter a pet battle with no available upgrades.
	Copyright (c) 2012-2018 Phanx <addons@phanx.net>. All rights reserved.
	https://www.wowinterface.com/downloads/info21998-PetBattleAutoForfeit.html
	https://www.curseforge.com/wow/addons/petbattleautoforfeit
	https://github.com/Phanx/PetBattleAutoForfeit
----------------------------------------------------------------------]]

local ADDON, private = ...
local L = private.L

------------------------------------------------------------------------

local db
PetBattleAutoForfeitDB = {
	enable = true,
	minLevelGain = 3,
	minQuality = 3, -- 1: poor, 2: common, 3: uncommon, 4: rare | shifted +1 vs values from GetItemInfo and in ITEM_QUALITY_COLORS
}

------------------------------------------------------------------------

local f = CreateFrame("Button", ADDON, UIParent)
f:SetAllPoints(true)
f:SetFrameStrata("FULLSCREEN_DIALOG")
f:RegisterForClicks("AnyUp")
f:Hide()

f.bg = f:CreateTexture(nil, "BACKGROUND")
f.bg:SetAllPoints(true)
f.bg:SetColorTexture(0, 0, 0, 0.5)

f.text1 = f:CreateFontString(nil, "OVERLAY", "SubZoneTextFont")
f.text1:SetPoint("BOTTOM", f, "CENTER", 0, 24)
f.text1:SetTextColor(1, 0.7, 0)
f.text1:SetText(L.NoUpgrades)

f.text2 = f:CreateFontString(nil, "OVERLAY", "PVPInfoTextFont")
f.text2:SetPoint("TOP", f, "CENTER", 0, 0)
f.text2:SetText(L.ClickForfeit)

f:SetScript("OnClick", function(self, button)
	if button == "LeftButton" then
		C_PetBattles.ForfeitGame()
	end
	self:Hide()
end)

------------------------------------------------------------------------

local LibPetJournal = LibStub("LibPetJournal-2.0")

local function IsUpgrade(i)
	local species = C_PetBattles.GetPetSpeciesID(LE_BATTLE_PET_ENEMY, i)
	local name, _, _, _, _, _, wild, _, _, _, obtainable = C_PetJournal.GetPetInfoBySpeciesID(species)
	if not wild or not obtainable then return false end

	local quality = C_PetBattles.GetBreedQuality(LE_BATTLE_PET_ENEMY, i)
	if quality < db.minQuality then return end

	local bestQuality, bestLevel = 0, 0
	for _, petID in LibPetJournal:IteratePetIDs() do
		local petSpecies, _, petLevel = C_PetJournal.GetPetInfoByPetID(petID)
		if petSpecies == species then
			local _, _, _, _, petQuality = C_PetJournal.GetPetStats(petID)
			if petQuality >= bestQuality then
				bestQuality = petQuality
				bestLevel = max(bestLevel, petLevel)
			end
		end
	end

	if quality > bestQuality then
		return true
	elseif quality == bestQuality then
		local level = C_PetBattles.GetLevel(LE_BATTLE_PET_ENEMY, i)
		if level >= 20 then
			level = level - 2
		elseif level >= 16 then
			level = level - 1
		end
		if (level - db.minLevelGain) > bestLevel then
			return true
		end
	end
end

f:RegisterEvent("PLAYER_LOGIN")
f:RegisterEvent("PET_BATTLE_OPENING_START")
f:SetScript("OnEvent", function(self, event)
	if not db then
		db = PetBattleAutoForfeitDB
		self:UnregisterEvent("PLAYER_LOGIN")
		-- Upgrade old settings
		if PBAF_ENABLE ~= nil then
			PetBattleAutoForfeitDB.enable = PBAF_ENABLE
			PBAF_ENABLE = nil
		end
		if PBAF_MIN_LEVEL_DIFF then
			PetBattleAutoForfeitDB.minLevelGain = PBAF_MIN_LEVEL_DIFF
			PBAF_MIN_LEVEL_DIFF = nil
		end
		if PBAF_MIN_QUALITY then
			PetBattleAutoForfeitDB.minQuality = PBAF_MIN_QUALITY
			PBAF_MIN_QUALITY = nil
		end
	end
	if db.enable and C_PetBattles.IsWildBattle() then
		for i = 1, C_PetBattles.GetNumPets(LE_BATTLE_PET_ENEMY) do
			if IsUpgrade(i) ~= nil then -- don't prompt vs non-capturable NPCs (IsUpgrade == false)
				return
			end
		end
		self:Show()
	end
end)

PetBattleFrame.BottomFrame.ForfeitButton:SetScript("OnClick", function(...)
	if IsShiftKeyDown() then
		C_PetBattles.ForfeitGame()
	else
		PetBattleForfeitButton_OnClick(...)
	end
end)

hooksecurefunc("PetBattleUnitFrame_UpdateDisplay", function(self)
	if not self.UpgradeIcon then
		local UpgradeIcon = self:CreateTexture(nil, "ARTWORK", nil, 3)
		UpgradeIcon:SetTexture("Interface\\ContainerFrame\\UI-Icon-QuestBang")
		UpgradeIcon:SetAllPoints(self.Icon)
		self.UpgradeIcon = UpgradeIcon
		if self.BorderAlive then
			self.BorderAlive:SetDrawLayer("ARTWORK", 4)
		end
	end
	local owner, i, wild = self.petOwner, self.petIndex, C_PetBattles.IsWildBattle()
	if i and wild and owner == LE_BATTLE_PET_ENEMY and i <= C_PetBattles.GetNumPets(owner) and IsUpgrade(i) then
		self.UpgradeIcon:Show()
	else
		self.UpgradeIcon:Hide()
	end
end)
